infra/build/functions/update_build_status.py

```diff
@@ -24,71 +24,138 @@
 import build_project
 import builds_status
 from datastore_entities import BuildsHistory
+from datastore_entities import LastSuccessfulBuild
 from datastore_entities import Project
 
 BADGE_DIR = 'badge_images'
 DESTINATION_BADGE_DIR = 'badges'
+MAX_BUILD_LOGS = 7
 
 
 class MissingBuildLogError(Exception):
   """Missing build log file in cloud storage."""
 
 
+def upload_status(data, status_filename):
+  """Upload json file to cloud storage."""
+  bucket = builds_status.get_storage_client().get_bucket(
+      builds_status.STATUS_BUCKET)
+  blob = bucket.blob(status_filename)
+  blob.cache_control = 'no-cache'
+  blob.upload_from_string(json.dumps(data), content_type='application/json')
+
+
+def sort_projects(projects, statuses):
```

Comment:
This is unnecessarily complex.

Just use a custom key function with the builtin Python sort() (which is guaranteed to be stable). then you don't need this `statuses` dict either.

i.e. something like: 

```python
def sort_projects(projects):
  def key_func(item):
    if not project['history']:
      return 2  # Order projects without history last.

     if project['history'][0]['success']:
       # Successful builds come second. 
       return 1

     # Build failures come first.
     return 0

  projects.sort(key=key_func)